interfaces/prompting/tags.go

```
@@ -0,0 +1,141 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2025 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package prompting
+
+import (
+	"fmt"
+	"math"
+	"sort"
+
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	prompting_errors "github.com/snapcore/snapd/interfaces/prompting/errors"
+	"github.com/snapcore/snapd/sandbox/apparmor/notify"
+	"github.com/snapcore/snapd/strutil"
+)
+
+var (
+	apparmorInterfaceForMetadataTag = apparmor.InterfaceForMetadataTag
+)
+
+// InterfaceFromTagsets returns the most specific interface which applies to
+// all of the given tagsets.
+//
+// Potential interfaces are identified by looking up whether a given tag is
+// registered in association with a particular snapd interface.
+//
+// If none of the given tags are registered with an interface, then returns
+// ErrNoInterfaceTags.
+//
+// If there are tags registered with an interface, but no interface applies to
+// every tagset, then returns an error.
+func InterfaceFromTagsets(tagsets notify.TagsetMap) (iface string, err error) {
```

Do we need all of this complexity? I'd much much prefer something that's deterministic and cannot be misused.

---

It is deterministic, it's just an annoying complex process, since:
1. We *need* to choose an interface which applies to every requested permission
2. We *want* to choose the most specific interface if somehow more than one interface applies to all requested permissions --- this is partially to make it deterministic, partially to make it semantically reflect what we want (e.g. the interface for which the specific rule applies, not some sort of broader interface which happens to cover the path)

But generally, we should only see one interface which applies to all requested permissions (or any, really). The "averages" just chooses the most specific one in the off-chance there are multiple. Same with the arbitrary tie-breaker.

Perhaps we could just return an error if there are more than one interfaces which apply to all permissions? The problem with that approach is we could end up with blocking requests in production when we could correctly handle them if we just choose one of the interfaces deterministically.

---

This has now been fully rewritten and simplified.